passes/cmds/partition.cc

```diff
@@ -0,0 +1,501 @@
+			if (!conn.second.is_wire() || !conn.second.as_wire()->port_input) {
+				for (auto i = 0; i < GetSize(conn.first); ++i) {
+					auto it = sigbit_to_edgenum.find(conn.second[i]);
+					if (it != sigbit_to_edgenum.end())
+						sigbit_to_edgenum.emplace(conn.first[i], it->second);
+					else
+						log_cmd_error("Cannot find driver for connection %s = %s\n", log_signal(conn.first[i]), log_signal(conn.second[i]));
+				}
+			}
+		}
+
+		//Fully populate edge_nodes, relying on a consistent iteration order over `module->cells()` to avoid storing
+		//a dict<RTLIL::Cell *, int> cell_to_nodenum:
+		node_ctr = 0;
+		for (auto cell : module->cells()) {
+			for (auto port : cell->connections()) {
+				if (port.first != ID::Y) {
+					for (auto i = 0; i < GetSize(port.second); ++i) {
+						auto sigbit = port.second[i];
+						auto it = sigbit_to_edgenum.find(sigbit);
+						if (it != sigbit_to_edgenum.end()) {
+							if (opt_verbose) log("Found driver hyperedge #%d for %s on node #%d\n", it->second, log_signal(sigbit), node_ctr);
+							edge_nodes[it->second].insert(node_ctr);
+						} else {
+							if (sigbit.wire == nullptr || !sigbit.wire->port_input)
+								log_cmd_error("Cannot find driver for net %s\n", log_signal(sigbit));
+						}
+					}
+				}
+			}
+			node_ctr++;
+		}
+	}
+
+	void write_hypergraph(const std::string &out_fname) {
+		//#edges #nodes 10
+		//[[node] foreach node in edge_nodes[i]] foreach 0 <= i < edge_ctr
+		//[weight(cell[i])] foreach 0 <= i < node_ctr
+		std::ofstream fout(out_fname);
+		if (!fout)
+			log_cmd_error("Could not open temporary file for writing.\n");
+
+		fout << edge_ctr << " " << node_ctr << " 10" << std::endl;
+		fout << "% edge definitions" << std::endl;
+		for (auto i = 0; i < edge_ctr; ++i) {
+			std::vector<int> nodes(edge_nodes[i].begin(), edge_nodes[i].end());
+			for (auto j = 0; j < GetSize(nodes); ++j) {
+				fout << nodes[j] + 1; //shift from 0-based indexing to 1-based indexing
+				if (j != GetSize(nodes) - 1)
+					fout << " ";
+			}
+			fout << std::endl;
+		}
+		fout << "% node weights" << std::endl;
+		for (auto i = 0; i < node_ctr; ++i) {
+			RTLIL::Cell *cell = *(module->cells().begin() + i);
+			fout << cell_type_costs.at(cell->type) << std::endl;
+		}
+		fout.close();
+	}
+
+	void call_solver(const std::string &problem_fname, const std::string &solution_fname) {
+		if (opt_k == -1 && opt_s == -1)
+			partitions = 2;
+		else if (opt_k != -1 && opt_k >= 2)
+			partitions = opt_k;
+		else if (opt_k != -1)
+			return; //nothing to do
+		else if (opt_s != -1) {
+			dict<IdString, double> cell_area;
+			statdata_t module_stats(design, module, false, cell_area, "gates");
+			unsigned long num_gates = 0;
+			for (auto &it : module_stats.num_cells_by_type)
+				num_gates += cell_type_costs[it.first] * it.second;
+			partitions = num_gates / opt_s;
+			if (partitions <= 1)
+				return; //nothing to do
+		}
+		const std::string partitions_str = opt_k != 1? stringf(" -k %d", partitions) : "";
```

Comment:
Is there any benefit in having `const` here, or does it just add visual noise?

Reply:
I suppose it's just visual noise.  Arguably anything beyond `auto` is visual noise...

Reply:
Sometimes you're writing out the type because you use it in list-initialization; and sometimes you're casting (or avoiding casting) the result of a function call, where the undefinedness of out-of-range numeric conversions makes the explicitness important. But yes.